File: Book Studio/AudioBookStudio.Models/Data/VideoTemplate.cs
namespace AudioBookStudio.Models.Data;

public record Rect(double X, double Y, double Width, double Height);

public record Location(double X, double Y);

public record Size(double Width, double Height);

public record Margin(double Top, double Right, double Bottom, double Left);

public class Transition
{
    public string Effect { get; set; } = "fade";
    public double Duration { get; set; } = 0.5;
}

public class Sizes
{
    public static Size YouTubeHD => new(1920, 1080);
    public static Size YouTubeShort => new(1080, 1920);
}

public class Locations
{
    public static Location TopLeft => new(0, 0);
    public static Location TopCenter => new(0.5, 0);
    public static Location TopRight => new(1, 0);
    public static Location MiddleLeft => new(0, 0.5);
    public static Location Center => new(0.5, 0.5);
    public static Location MiddleRight => new(1, 0.5);
    public static Location BottomLeft => new(0, 1);
    public static Location BottomCenter => new(0.5, 1);
    public static Location BottomRight => new(1, 1);
}

public class FontNames
{
    public static string Arial => "Arial";
    public static string YaHei => "Microsoft YaHei";
}

public class Weights
{
    public static string Normal => "Normal";
    public static string Bold => "Bold";
    public static string Light => "Light";
}

public class Colors
{
    public static string White { get; set; } = "#FFFFFF";
    public static string Black { get; set; } = "#000000";
}

public class Margins
{
    public static Margin None => new(0, 0, 0, 0);
    public static Margin Small => new(5, 5, 5, 5);
    public static Margin Medium => new(10, 10, 10, 10);
    public static Margin Large => new(20, 20, 20, 20);
}

public class VideoTemplate : Entity
{
    public required string Name { get; set; }
    public string? Description { get; set; }
    public string? Thumbnail { get; set; }
    public Size Resolution { get; set; } = Sizes.YouTubeHD;
    public DateTime? DateCreated { get; set; }
    public DateTime? DateUpdated { get; set; }
    public List<VideoTrack> Tracks { get; set; } = [];
    public List<TemplateProperty> Properties { get; set; } = [];
}

File: Book Studio/matrixbook.client/src/app/models/video-template.ts


export interface Rect {
    x: number;
    y: number;
    width: number;
    height: number;
}

export interface Location {
    x: number;
    y: number;
}

export interface Size {
    width: number;
    height: number;
}

export interface Margin {
    top: number;
    right: number;
    bottom: number;
    left: number;
}

export interface Transition {
    effect: string;
    duration: number;
}

export const PropertyTypes = {
    text: 'Text',
    resource: 'Resource'
} as const;
export type PropertyType = (typeof PropertyTypes)[keyof typeof PropertyTypes];


export interface TemplateProperty {
    id: string;
    name: string;
    value: string;
    tag?: string;
    type: PropertyType;
}

export interface VideoTemplate {
    id: string;
    name: string;
    description: string;
    thumbnail: string;
    resolution: Size;
    dateCreated: Date;
    dateUpdated: Date;
    tracks: VideoTrack[];
    properties?: TemplateProperty[];
}

export interface VideoTrack {
    id: string;
    name: string;
    elements: MediaElement[];
}

export interface LayoutOptions {
    location?: Location;
    size?: Size;
    align?: Align;
    margin?: Margin;
}

export interface FontOptions {
    family: string;
    fontSize: number;
    weight: 'normal' | 'bold' | 'italic';
    color?: string;
    backgroundColor?: string;
    shadow?: boolean;
}

export const Alignments = {
    top: 'top',
    center: 'center',
    bottom: 'bottom',
    left: 'left',
    right: 'right',
    topLeft: 'topLeft',
    topRight: 'topRight',
    bottomLeft: 'bottomLeft',
    bottomRight: 'bottomRight',
    middleLeft: 'middleLeft',
    middleRight: 'middleRight'
} as const;
export type Align = (typeof Alignments)[keyof typeof Alignments];

export const DurationTypes = {
    fixed: 'Fixed',
    auto: 'Auto',
    full: 'Full'
} as const;
export type DurationType = (typeof DurationTypes)[keyof typeof DurationTypes];

export const StartTypes = {
    absolute: 'Absolute',
    afterPrevious: 'After Previous',
    beforeEnd: 'Before End'
} as const;
export type StartType = (typeof StartTypes)[keyof typeof StartTypes];

export const MediaTypes = {
    audio: 'audio',
    image: 'image',
    video: 'video',
    text: 'text',
    combine: 'combine'
} as const;
export type MediaType = (typeof MediaTypes)[keyof typeof MediaTypes];

export interface MediaElement {
    type: MediaType;
    name: string;
    start: number;
    duration: number;
    fadeIn: number;
    fadeOut: number;
    alpha: number;
    transition?: Transition;
    durationType?: DurationType;
    startType?: StartType;
    layout?: LayoutOptions;
    inputSource?: string;
    textContent?: string;
    children?: MediaElement[];
}
